providers/dns/liquidweb/liquidweb.go

```diff
@@ -0,0 +1,139 @@
+type Config struct {
+	URL                string
+	Username           string
+	Password           string
+	Zone               string
+	Timeout            time.Duration
+	PropagationTimeout time.Duration
+}
+
+// NewDefaultConfig returns a default configuration for the DNSProvider
+func NewDefaultConfig() *Config {
+	config := &Config{
+		URL:                env.GetOrDefaultString("LW_URL", ""),
+		Username:           env.GetOrDefaultString("LW_USERNAME", ""),
+		Password:           env.GetOrDefaultString("LW_PASSWORD", ""),
+		Zone:               env.GetOrDefaultString("LW_ZONE", ""),
+		Timeout:            time.Second * time.Duration(env.GetOrDefaultInt("LW_TIMEOUT", 60)),
+		PropagationTimeout: time.Minute * 5,
+	}
+
+	return config
+}
+
+// DNSProvider is an implementation of the acme.ChallengeProvider interface
+// that uses Liquid Web's REST API to manage TXT records for a domain.
+type DNSProvider struct {
+	config      *Config
+	recordIDs   map[string]int
+	recordIDsMu sync.Mutex
+	client      *lwclient.API
+}
+
+// NewDNSProvider returns a DNSProvider instance configured for Liquid Web.
+func NewDNSProvider() (*DNSProvider, error) {
+	config := NewDefaultConfig()
+
+	return NewDNSProviderConfig(config)
+}
+
+// NewDNSProviderConfig return a DNSProvider instance configured for Liquid Web.
+func NewDNSProviderConfig(config *Config) (*DNSProvider, error) {
+	if config == nil {
+		return nil, errors.New("the configuration of the DNS provider is nil")
+	}
+
+	if config.URL == "" {
+		return nil, fmt.Errorf("liquidweb: url is missing")
+	}
+
+	if config.Username == "" {
+		return nil, fmt.Errorf("liquidweb: username is missing")
+	}
+
+	if config.Password == "" {
+		return nil, fmt.Errorf("liquidweb: password is missing")
+	}
+
+	// Initial new LW go client.
+	lwAPI, err := lwclient.NewAPI(config.Username, config.Password, config.URL, int(config.Timeout.Seconds()))
+
+	if err != nil {
+		log.Fatalf("Could not create Liquid Web API client: %v", err)
+	}
+
+	return &DNSProvider{
+		config:    config,
+		recordIDs: make(map[string]int),
+		client:    lwAPI,
+	}, nil
+}
+
+// Timeout returns the timeout and interval to use when checking for DNS propagation.
+// Adjusting here to cope with spikes in propagation times.
+func (d *DNSProvider) Timeout() (timeout, interval time.Duration) {
+	return d.config.Timeout, d.config.PropagationTimeout
+}
+
+// Present creates a TXT record using the specified parameters
+func (d *DNSProvider) Present(domain, token, keyAuth string) error {
+	fqdn, value := dns01.GetRecord(domain, keyAuth)
+	params := &network.DNSRecordParams{
+		Name:  fqdn[0 : len(fqdn)-1],
+		RData: strconv.Quote(value),
+		Type:  "TXT",
+		Zone:  d.config.Zone,
+	}
+
+	dnsEntry, err := d.client.NetworkDNS.Create(params)
+	if err != nil {
+		return fmt.Errorf("could not create TXT record: %v", err)
+	}
+
+	d.recordIDsMu.Lock()
+	d.recordIDs[token] = int(dnsEntry.ID)
+	d.recordIDsMu.Unlock()
```

Comment:
could you explains why you are using a lock?

Reply:
I started this work off the back of the Digital Ocean DNS provider. I went through the code and couldn't find anything indicating locks would be necessary and have removed them. Nice catch.